### ISSUE TITLE
PSpecContainer SWMR mode fails when repeatedly opening file in 'rw' mode

### ISSUE THREAD

**Issue by nkern** [HERA-Team/hera_pspec#214]
I'm discovering an odd bug when opening a `PSpecContainer` in transactional 'rw' mode (using SWMR). Here is a working example from the `test_grouping` file
```
import numpy as np
import os
from hera_pspec.data import DATA_PATH
from hera_pspec import uvpspec, conversions, parameter, pspecbeam, pspecdata, testing, utils
from hera_pspec import uvpspec_utils as uvputils
from hera_pspec import grouping, container
from pyuvdata import UVData
from hera_cal import redcal
import copy

visfile = os.path.join(DATA_PATH, "zen.even.xx.LST.1.28828.uvOCRSA")
beamfile = os.path.join(DATA_PATH, "HERA_NF_dipole_power.beamfits")
cosmo = conversions.Cosmo_Conversions()
beam = pspecbeam.PSpecBeamUV(beamfile, cosmo=cosmo)
uvd = UVData()
uvd.read_miriad(visfile)
ap, a = uvd.get_ENU_antpos(pick_data_ants=True)
reds = redcal.get_pos_reds(dict(zip(a, ap)), bl_error_tol=1.0)[:3]
uvp = testing.uvpspec_from_data(uvd, reds, spw_ranges=[(50, 100)], beam=beam, cosmo=cosmo)
if os.path.exists("ex.h5"):
    os.remove("ex.h5")
psc = container.PSpecContainer("ex.h5", mode='rw', keep_open=False)

# set a pspec in SWMR 'rw' mode, this works fine
psc.set_pspec("grp1", "uvp", uvp, overwrite=True)

# now do it multiple times with a new spectra name
for i in range(10):
  psc.set_pspec("grp1", "uvp{}".format(i), uvp, overwrite=True)
```
and we get this traceback after only 5 or 6 loops:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-1-c361d82ae83f> in <module>()
     27 # now do it multiple times with a new spectra name
     28 for i in range(10):
---> 29   psc.set_pspec("grp1", "uvp{}".format(i), uvp, overwrite=True)

/Users/nkern/Software/anaconda2/lib/python2.7/site-packages/hera_pspec-0.1.0-py2.7.egg/hera_pspec/container.pyc in wrapper(*args, **kwargs)
     28             if not psc.keep_open:
     29                 psc._close()
---> 30             raise err
     31 
     32         # Close HDF5 file if necessary

ValueError: Unable to create group (Parent entry isn't pinned)
```
I've tried to track this down and I've found very little on the interwebs: but it does seem like a fundamental issue with HDF5, not with `h5py`. For reference I'm running Python 2.7.3, h5py 2.8.0 and HDF5 1.10.2.

What is interesting is that if I edit the `PSpecContainer._open` function and set Line 135 to `self.data.swmr_mode = False` such that SWMR is never turned on for the writer task, this error goes away, making me think that its a bug with HDF5 SWMR.

My work around is to allow an option in `PSpecContainer._open` to not turn on `swmr` if `mode='rw'`, which before was done always by default. #212 

@plaplant @philbull

**Comment by plaplant** [comment]
If I had to bet, this is related to an issue that I raised when I reviewed PR #193. According to the SWMR spec, it is _forbidden_ to add a new group or dataset to a file while it is in SWMR mode. To me, the PR was a non-starter as it was, and it should not have been merged without addressing that fundamental flaw. Apologies if my review did not make that clear.

This is not a bug in the library, but rather a well-known, well-documented limitation. One potential workaround might be to generate all necessary datasets on first touch with `swmr_mode=False`, and then re-open the file in SWMR mode to actually populate the datasets with data. I am not knowledgable enough about the typical access patters of the pspec code to know if this is viable, or if the number and size of datasets is known at the time of file creation. Barring a workaround like this, SWMR may not be viable for the pspec containers as the code is currently structured.

**Comment by nkern** [comment]
Hey, apologies if you weren't ok with the merge, it wasn't clear to me that it was a non-starter for you, and also what exactly wanted as a fix: I tried to address all of the concrete things you asked for.

I actually don't know if what you raise is the issue, because as I show in the example above:
1) there are never more than one concurrent task in SWMR-writer mode, only one at a time, and
 2) I can indeed add a group (and also a few), its just that for some reason when I add more than, say 5, it errors..

**Comment by nkern** [comment]
To followup on this @plaplant, this could still be the issue you are talking about but just that the error isn't a hard-and-fast rule, and isn't caught up-front by HDF5 of `h5py`. The HDF5 SWMR guide states

> "It is the user’s responsibility to follow the SWMR scope since the HDF5 Library does not enforce the scope."

so it may just be that one can actually write a new dataset with SWMR but we aren't supposed to b/c of some reason they aren't totally up front about, and cause weird bugs like the one we are seeing.

Regardless, I don't think the answer is to not allow SWMR with our containers, we just need to use them in a way that is SWMR compliant, so I don't think we would have changed anything from #193 anyways other than some print warnings and docstring editing. I'll put that into the PR i'm working on #212

**Comment by philbull** [comment]
Hi @plaplant, sorry I got the wrong end of the stick with your review comments. I think the merge was safe enough, as it solves a bug with concurrent read access, but I also agree that we should disable swmr for rw mode. AFAIK there is no data loss risk here as the HDF5 library is doing the right thing and preventing invalid operations, so this is not super-critical.

This does seem to be quite an annoying limitation of HDF5, and I think any concurrent-write capability we come up with will be hacky as a result. (I did start implementing a queue/lock system at one point, but demurred as it became increasingly hacky).

After thinking about this quite a bit, and searching for other concurrent-write file formats, I've come to the conclusion that the most sensible solution is to just use the filesystem! There are fs-based solutions like ExDir (https://exdir.readthedocs.io/en/latest/), which would be a drop-in replacement for HDF5, but I don't know if we even need that.

So, is there any functionality of `PSpecContainer` that we'd miss if we switched to just dumping `UVPSpec` objects into directories as individual `uvh5` files?

**Comment by philbull** [comment]
> > "It is the user’s responsibility to follow the SWMR scope since the HDF5 Library does not enforce the scope."

This is horrendous. Definitely not what we want in a storage library!

> so it may just be that one can actually write a new dataset with SWMR but we aren't supposed to b/c of some reason they aren't totally up front about, and doing so may cause weird bugs like the one we are seeing.

I have seen this mentioned in the docs. I think the risk is that you can corrupt the HDF5 file.

**Comment by nkern** [comment]
I think @plaplant suggested a good idea, which is that any time we call `PSpecContainer.set_pspec` we check if it will be writing a new group or dataset and raise an error if so. I *think* that should cover our bases.

Also, @philbull, I think turning off `swmr` for `'rw'` mode actually eliminates the entire purpose of adding `swmr` in the first place: we can already do concurrent read across multiple processes, so the only benefit of `swmr` (to my knowledge) is to allow concurrent read access in the presence of a writer process.

**Comment by nkern** [comment]
Addressed by #212
